cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -748,18 +748,77 @@
 (decl put_in_reg_mem (Value) RegMem)
 (extern constructor put_in_reg_mem put_in_reg_mem)
 
+;; Addressing modes.
+
 (type SyntheticAmode extern (enum))
 
 (decl synthetic_amode_to_reg_mem (SyntheticAmode) RegMem)
 (extern constructor synthetic_amode_to_reg_mem synthetic_amode_to_reg_mem)
 
+(decl amode_to_synthetic_amode (Amode) SyntheticAmode)
+(extern constructor amode_to_synthetic_amode amode_to_synthetic_amode)
+
 (type Amode extern (enum))
 
+(decl amode_imm_reg (u32 Gpr) Amode)
+(extern constructor amode_imm_reg amode_imm_reg)
+
 (decl amode_imm_reg_reg_shift (u32 Gpr Gpr u8) Amode)
 (extern constructor amode_imm_reg_reg_shift amode_imm_reg_reg_shift)
 
-(decl amode_to_synthetic_amode (Amode) SyntheticAmode)
-(extern constructor amode_to_synthetic_amode amode_to_synthetic_amode)
+;; A helper to check if a shift amount (the `Value`) is both constant and
+;; less-than or equal to 3; this is needed since x64 can only shift addresses
+;; using two bits.
+(decl small_const_shift (u8) Value)
+(extern extractor small_const_shift small_const_shift)
+
+;; A helper to check if a constant offset value will fit into the 32-bit maximum
+;; allowed by the x64 SIB encoding. (TODO re-document, this might not be the
+;; correct explanation)
+(decl low32_will_sign_extend_to_64 (u32) Imm64)
+(extern extractor low32_will_sign_extend_to_64 low32_will_sign_extend_to_64)
+
+;; To generate an address for a memory access, we can pattern-match various CLIF
+;; sub-trees to x64's complex addressing modes (`Amode`). In pseudo-code:
+;;
+;; if address matches iadd(a, b): if either a or b: matches (ishl c with shift
+;;   amount <= 3): amode(base + offset + c >> amount) matches (iconst c where c
+;;     will sign-extend to 64 bits): amode(base + (c + offset)) matches (uextend
+;;       (iconst c) where c will sign-extend to 64 bits): amode(base + (c +
+;;     offset)) else: amode(a + b + offset >> 0) else: amode(base + offset)
+;;
+;; The rules for `to_amode` correspond to a subset of the possible addressing
+;; modes available by tweaking the SIB byte, the MOD bits, and the size of the
+;; displacement (i.e., offset). More information is available in Intel's
+;; Software Developer's Manual, volume 2, section 2.1.5, "Addressing-Mode
+;; Encoding of ModR/M and SIB Bytes."
+;;
+;; TODO should this also handle `ishl_imm`?
```

Comment:
I think `ishl_imm` is converted during legalization to `ishl` + `iconst` so we can probably remove this TODO.